examples/go/vm-reconfig-via-tag/handler/handler.go

```diff
@@ -0,0 +1,564 @@
+func validateConfig(cfg vcConfig) error {
+	reqFields := map[string]string{
+		"vcenter server":   cfg.VCenter.Server,
+		"vcenter user":     cfg.VCenter.User,
+		"vcenter password": cfg.VCenter.Password,
+	}
+
+	// Multiple fields may be missing, but err on the first encountered.
+	for k, v := range reqFields {
+		if v == "" {
+			return errors.New("required field(s) missing, including " + k)
+		}
+	}
+
+	return nil
+}
+
+// eventVmMOR parses the cloud event and gets the triggering VM's managed object
+// reference.
+func eventVmMOR(req []byte) (*types.ManagedObjectReference, error) {
+	var ce cloudEvent
+	var mor types.ManagedObjectReference
+
+	err := json.Unmarshal(req, &ce)
+	if err != nil {
+		return nil, fmt.Errorf("parse request: %w", err)
+	}
+
+	if ce.Data.Vm == nil || ce.Data.Vm.Vm.Value == "" {
+		return nil, errors.New("event object is not a VM")
+	}
+
+	// Fill information in the request into a govmomi type.
+	mor.Type = ce.Data.Vm.Vm.Type
+	mor.Value = ce.Data.Vm.Vm.Value
+
+	if debug {
+		log.Println("Cloud event parsed and validated.")
+	}
+
+	return &mor, nil
+}
+
+// newClient connects to vSphere govmomi API
+func newClient(ctx context.Context, cfg *vcConfig) (*vsClient, error) {
+	u := url.URL{
+		Scheme: "https",
+		Host:   cfg.VCenter.Server,
+		Path:   "sdk",
+	}
+
+	u.User = url.UserPassword(cfg.VCenter.User, cfg.VCenter.Password)
+	insecure := cfg.VCenter.Insecure
+
+	gc, err := govmomi.NewClient(ctx, &u, insecure)
+	if err != nil {
+		return nil, fmt.Errorf("connect to vSphere API: %w", err)
+	}
+
+	rc := rest.NewClient(gc.Client)
+	tm := tags.NewManager(rc)
+
+	vsc := vsClient{
+		govmomi: gc,
+		rest:    rc,
+		tagMgr:  tm,
+	}
+
+	err = vsc.rest.Login(ctx, u.User)
+	if err != nil {
+		return nil, fmt.Errorf("log into REST API: %w", err)
+	}
+
+	if debug {
+		session, _ := vsc.govmomi.SessionManager.UserSession(ctx)
+
+		log.Printf(
+			"New connection to vSphere established. Govmomi SessionID: %s, REST SessionID: %s",
+			session.Key,
+			vsc.rest.SessionID(),
+		)
+	}
+
+	return &vsc, nil
+}
+
+func handleSignal(ctx context.Context, vsc *vsClient) {
+	sigCh := make(chan os.Signal, 2)
+
+	signal.Notify(sigCh, syscall.SIGTERM, os.Interrupt)
+
+	if debug {
+		log.Println("Started signal handling.")
+	}
+
+	s := <-sigCh
+
+	err := vsc.govmomi.Logout(ctx)
+	if err != nil {
+		log.Printf("log out of Govmomi: %v", err)
+	}
+
+	err = vsc.rest.Logout(ctx)
+	if err != nil {
+		log.Printf("log out of VAPI: %v", err)
+	}
+
+	if debug {
+		log.Printf("got signal: %v, logged out of vSphere APIs", s)
+	}
+}
+
+func (c *vsClient) attachedTags(ctx context.Context, mor types.ManagedObjectReference) ([]tags.Tag, error) {
+	// Tag objects attached to the VM.
+	tagObjs, err := c.tagMgr.GetAttachedTags(ctx, mor)
+	if err != nil {
+		return []tags.Tag{}, fmt.Errorf("get attached tags: %w", err)
+	}
+
+	if debug {
+		log.Println("Retrieved", len(tagObjs), "tag(s).")
```

Comment:
Missing spaces between strings.

Reply:
`Println` adds spaces in between the arguments, so no spaces needed in the strings. Here's a Go Playground example: https://play.golang.org/p/h8GtadI0q9K

Reply:
Ah, did not know that since formatting print statements is typically done via `Printf` handlers for clarity. Nevermind.